app/presenters/tree_builder_vms_and_templates.rb

```diff
@@ -50,25 +51,43 @@ def reparent_hidden_folders(tree)
     end
   end
 
-  def prune_rbac(tree)
-    allowed_vm_ids = Set.new(Rbac.filtered(@root_ems.vms).pluck(:id))
-
-    prune_filtered_vms(tree, allowed_vm_ids)
-    prune_empty_folders(tree)
+  def prune_rbac(tree, display_vms)
+    if display_vms
+      allowed_vm_ids = Set.new(Rbac.filtered(@root_ems.vms).pluck(:id))
+      prune_folders_via_vms(tree, allowed_vm_ids)
+    else
+      allowed_vms = Rbac.filtered(@root_ems.vms) # this is a sub-query
+      vm_relations = Relationship.where(:resource     => allowed_vms,
+                                        :relationship => "ems_metadata")
+                                 .select(:ancestry).distinct # bigger sub-query
+                                 .map(&:parent_id)
+      allowed_folder_ids = Relationship.where(:id => vm_relations).pluck(:resource_id)
+      prune_folders_via_folders(tree, allowed_folder_ids)
+    end
   end
 
-  def prune_filtered_vms(tree, allowed_vm_ids)
+  def prune_folders_via_vms(tree, allowed_vm_ids)
     tree.reject! do |object, children|
-      prune_filtered_vms(children, allowed_vm_ids)
-      object.kind_of?(VmOrTemplate) && !allowed_vm_ids.include?(object.id)
+      prune_folders_via_vms(children, allowed_vm_ids)
+      if object.kind_of?(VmOrTemplate)
+        !allowed_vm_ids.include?(object.id)
+      elsif object.kind_of?(EmsFolder)
+        children.empty?
+      end
     end
   end
 
-  def prune_empty_folders(tree)
-    tree.reject! do |object, children|
-      prune_empty_folders(children)
-      object.kind_of?(EmsFolder) && children.empty?
+  def prune_folders_via_folders(tree, allowed_folder_ids)
```

Comment:
the old approach is to prune vms based upon rbac and then prune folders that have no vm children (vm is not in the array) or folder children

the new approach is to get the list of folders that have vms, and then prune folders that have no vm children (folder is not in the list) or folder children